Adds repeatCharacterLimit option

mleanos opened the pull request:
Added repeatCharacterLimit option to restrict repetitive characters to
the specified amount. This option is disabled by default.

mleanos commented:
Related to #4 

I wasn't sure what to do here... 
https://github.com/brendanashworth/generate-password/pull/5/files#diff-d429632d5f0874d417c9e8e8cc0a3c5dR21

By requiring the repeatCharacterLimit option to be at least 2, if this option was provided, we can avoid any serious performance issues. However, it could be left to the user of this package to decide. They should understand the implications if they provided `repeatCharacterLimit: 1`. IMO, this wouldn't be a likely use case.

mleanos commented:
I just realized there's an issue with this approach.

It's possible to remove 3 repeating characters from the generated password, and the result having 3 or more repeating characters...

For example:
"f3dirssuuus48cDd343Dspqw" will become "f3dirsss48cDd343Dspqw"

mleanos commented:
This should handle the issue described in my last comment. 

I'm not sure how else this could be handled. 
https://github.com/brendanashworth/generate-password/pull/5/files#diff-d429632d5f0874d417c9e8e8cc0a3c5dR51

brendanashworth commented on src/generate.js:
```diff
@@ -17,11 +17,15 @@ self.generate = function(options) {
 	if (!options.hasOwnProperty('excludeSimilarCharacters')) options.excludeSimilarCharacters = false;
 
+	// check if must limit the number of repeating characters
+	if (!options.hasOwnProperty('repeatCharactersLimit') || typeof options.repeatCharactersLimit !== 'number' || options.repeatCharactersLimit < 2) options.repeatCharactersLimit = undefined;
```
You don't need to sanitize the passed option here (checking for number and value). Also, a repeat character limit of 0 seems to make the most sense, rather than undefined for no limit.

brendanashworth replied:
Sorry if I was unclear, the `typeof options.xxx !== 'number'` should be unnecessary :) type checking might be a good idea but there is no precedence to do it already. Perhaps another PR?

mleanos replied:
I just saw this new line comment. I'll update to conform to your property checks. My thinking was that, I wouldn't want the method to "blow up" if the user had provided a non-number value for this option. Perhaps another PR would be more appropriate.

mleanos commented:
@brendanashworth I addressed your line comment, and fixed a mistake in the `repeatingCharacters` regexp logic.

How do you like this approach overall? If you're fine with merging this, I can squash my commits beforehand.

mleanos commented:
I didn't see your last line comment until tonight. I changed the option check to conform to the rest of the options. I also squashed. 

WDYT?

brendanashworth commented on src/generate.js:
```diff
@@ -16,12 +16,14 @@ self.generate = function(options) {
 		symbols = '!@#$%^&*()+_-=}{[]|:;"/?.><,`~',
-		similarCharacters = /[ilLI|`oO0]/g;
+		similarCharacters = /[ilLI|`oO0]/g,
+		repeatingCharacters = options.repeatCharactersLimit ? new RegExp('(.)\\1{' + (options.repeatCharactersLimit) + ',}', 'g') : undefined;
```
Do you think it'd be better to put `options.repeatCharactersLimit > 1` here? Or should we throw on `1`?

mleanos replied:
Do you mean, if a user passes in `1` as the repeatCharactersLimit option, then throw an error?

It's been a while since I worked on this. I'll re-familiarize myself with it again, to have a better understanding of passing in `0` & `1` with this option. I forgot the conclusions I came up with on those use cases.

brendanashworth replied:
Yes, precisely. You commented earlier about it, if it helps you :)

> By requiring the repeatCharacterLimit option to be at least 2, if this option was provided, we can avoid any serious performance issues. However, it could be left to the user of this package to decide. They should understand the implications if they provided repeatCharacterLimit: 1. IMO, this wouldn't be a likely use case.

brendanashworth commented:
@mleanos sorry about dropping the ball here. I must have forgotten about this PR — I like where you're going with it, it looks great. If you wouldn't mind adding a test, I can merge this in pretty soon.

mleanos commented:
I'll spend some time on this over the next couple of days. And no worries. I can understand how this fell under the radar.

brendanashworth commented:
Ping @mleanos this requires a little bit of work, are you still up to implementing this?

brendanashworth commented:
I'm going to close this, feel free to reopen if you'd like to continue this.
